### PR TITLE
Create feature flag library for pilot

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
fixes: #8960

/cc @rshriram

### CODE REVIEW

File: pilot/cmd/pilot-agent/main.go [hunk 1]
```diff
@@ -98,7 +99,7 @@ var (
 			// set values from registry platform
 			if len(role.IPAddress) == 0 {
 				if registry == serviceregistry.KubernetesRegistry {
-					role.IPAddress = os.Getenv("INSTANCE_IP")
+					role.IPAddress = pilot.ProxyIPAddress
```
Comment (rshriram) [inline pilot/cmd/pilot-agent/main.go]: These can stay in current place. It’s not a feature flag. It’s just environment specific property for local Ip. Same goes for other env from the kubernetes downward Api

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -938,7 +939,7 @@ func (s *Server) initGrpcServer() {
 
 // The secure grpc will start when the credentials are found.
 func (s *Server) secureGrpcStart(listener net.Listener) {
-	certDir := os.Getenv("PILOT_CERT_DIR")
+	certDir := pilot.CertDir
```
Comment (rshriram) [inline pilot/pkg/bootstrap/server.go]: Nice

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 3]
```diff
@@ -129,7 +128,7 @@ func (s *DiscoveryServer) Register(rpcs *grpc.Server) {
 // ( will be removed after change detection is implemented, to double check all changes are
 // captured)
 func (s *DiscoveryServer) periodicRefresh() {
-	envOverride := os.Getenv("V2_REFRESH")
+	envOverride := pilot.RefreshDuration
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/discovery.go]: Shouldn’t this be a function as it gets evaluated in every call

File: pilot/pkg/serviceregistry/kube/client.go [hunk 4]
```diff
@@ -37,7 +39,7 @@ import (
 func ResolveConfig(kubeconfig string) (string, error) {
 	// Consistency with kubectl
 	if kubeconfig == "" {
-		kubeconfig = os.Getenv("KUBECONFIG")
+		kubeconfig = pilot.KubeConfig
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/client.go]: I think we should get rid of this function ?

File: pilot/cmd/pilot-agent/main.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/main.go]: make sense

File: pilot/pkg/proxy/envoy/v2/discovery.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: Ic, you mean by calling function we can dynamically update this env value.

File: pilot/pkg/serviceregistry/kube/client.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/client.go]: Indeed, this is only used by e2e.

File: pilot/pkg/proxy/envoy/v2/discovery.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: After another look, I find this func `periodicRefresh` is called only once. Maybe we can do env reload as a follow up .

**Comment by ozevren** [comment]
Why is this checked in istio/pkg? This seems awfully Pilot specific.

**Comment by rshriram** [comment]
so that others can use similar mechanisms..

**Pull request merged**